### PR TITLE
docs: document proxy support

### LINKED ISSUE #411: chore: document and test HTTP_PROXY support
# Problem

We don't have any documentation on using HTTP_PROXY, HTTPS_PROXY, NO_PROXY, or our own special ENVBUILDER_GIT_HTTP_PROXY_URL environment variables.

We should document this and add integration tests to validate their support.

### PR DESCRIPTION
**Pull request by SasSwart**
This PR closes #411

### CODE REVIEW

**Comment by SasSwart** [comment]
Still thinking about an integration test for this, to ensure that it always works. Personally, I'm happy to do that as a second PR. Keen to hear from reviewers though.

File: docs/proxy.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+Envbuilder can be used from behind transparent TLS proxies that would normally risk interrupting TLS verification.
+
+A summary of how to configure Envbuilder to run behind an HTTPS proxy is provided in the next session. Thereafter an illustrative example is provided that can be followed to prove the concept from first principles before applying it in production.
+
+## Summary
+To use Envbuilder behind a proxy that requires a custom certificate, the following configuration needs to be applied. Note that if you are using Envbuilder in conjunction with Coder, these variables should be set on the envbuilder container itself, not on the Coder agent.
+
+To configure Envbuilder to route HTTP traffic for git and the container registry to the proxy do at least one of the following:
+* Set the `https_proxy=https://host:port` environment variable for the envbuilder container. This will also proxy traffic for other programs in the container that respect `https_proxy`. If you do set it, look into the related `http_proxy` and `no_proxy` environment variables.
```
Comment (johnstcn) [inline docs/proxy.md:11]: Maybe add a note about how capitalization (infuriatingly) matters for this. 

Ref: https://about.gitlab.com/blog/2021/01/27/we-need-to-talk-no-proxy/#http_proxy-and-https_proxy

File: docs/proxy.md [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+![Proxied requests](./img/proxy.png)
+
+### Cleanup
+Once the demonstration has concluded, cleanup the artifacts that were used in the process:
+```bash
+docker stop mitmproxy
+rm -r certs/
+```
+
+You may require
```
Comment (johnstcn) [inline docs/proxy.md:128]: Unfinished thought?

File: docs/proxy.md [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+```bash
+exit
+```
+
+Let's now temporarily break Envbuilder by introducing a transparent TLS intercepting proxy. To do this, we'll use [mitmproxy](https://mitmproxy.org/). Start mitmproxy in a container, by running the following:
+```bash
+docker run --rm -d --user $(id -u):$(id -g) --name mitmproxy -v ./certs:/home/mitmproxy/.mitmproxy -p 8080:8080 -p 127.0.0.1:8081:8081 mitmproxy/mitmproxy mitmweb --web-host 0.0.0.0 --set http2=false
+```
+
+Notice that we disable HTTP2 on mitmproxy. This is because Envbuilder and mitmproxy do not seem to be able to negotiate which version of HTTP to use. mitmproxy seems to interpret Envbuilder's HTTP1.1 request as an HTTP2 request and then fails to find the expected HTTP2 preamble (because there is no HTTP2 preamble in an HTTP1.1 request). If your production proxy exhibits this behavior, please file a GitHub issue.
```
Comment (johnstcn) [inline docs/proxy.md:49]: nit: "seems to" dilutes the authority of this section a bit. Maybe re-word?

**Review by johnstcn — APPROVED** [review]
Nice work! Just a couple of nits. I especially like the demonstration you provided, it really aids understanding.

**Comment by johnstcn** [comment]
> Still thinking about an integration test for this, to ensure that it always works. I'm happy to do that as a second PR. Keen to hear from reviewers though.

Separate PR is fine 👍 but always a fan of an integration test for this sort of functionality.

File: docs/proxy.md [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+
+To configure Envbuilder to route HTTP traffic for git and the container registry to the proxy do at least one of the following:
+* Set the `https_proxy=https://host:port` environment variable for the envbuilder container. This will also proxy traffic for other programs in the container that respect `https_proxy`. If you do set it, look into the related `http_proxy` and `no_proxy` environment variables.
+* Set the `ENVBUILDER_GIT_HTTP_PROXY_URL=https://host:port` environment variable. This will specifically route traffic for Envbuilder's git and container registry without affecting other programs inside the container.
+
+Once traffic is routed to the proxy, you will need to install the proxy's CA certificate into Envbuilder. To do this, you can do one of the following:
+* Set `ENVBUILDER_SSL_CERT_BASE64=...` for Envbuilder to the base64 encoded value of your proxy's CA certificate. This will only apply to Envbuilder. Other TLS connections within your container will not be aware of this certificate.
```
Comment (mafredri) [inline docs/proxy.md:15]:
```suggestion
* Set `ENVBUILDER_SSL_CERT_BASE64=...` to the base64 encoded value of your proxy's CA certificate. This will only apply to Envbuilder. Other TLS connections within your container will not be aware of this certificate.
```
Nit: Repeat

File: docs/proxy.md [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+* Mount the certificate file into the Envbuilder container and then set `ENVBUILDER_SSL_CERT_FILE=/path/to/cert.pem`.
+* Mount a directory containing all relevant CA certificates into the Envbuilder container and then set `ENVBUILDER_SSL_CERT_DIR=/path/to/certs/`.
+
+## Demonstration
+Envbuilder clones a repository that contains your devcontainer.json and optional Dockerfile so that it can build your container. If the clone is done using HTTPS, then TLS verification will have to succeed, or be disabled. If a transparent HTTPS proxy is present, TLS verification will fail unless Envbuilder trusts the certificate used by the transparent proxy. Therefore, we need to tell Envbuilder how to trust your transparent proxy.
```
Comment (mafredri) [inline docs/proxy.md:20]:
```suggestion
Envbuilder clones a repository that contains your `devcontainer.json` and optional `Dockerfile` so that it can build your container. If the clone is done using HTTPS, then TLS verification must succeed or be disabled. If a transparent HTTPS proxy is present, TLS verification will fail unless Envbuilder trusts the proxy’s certificate. Therefore, we need to configure Envbuilder to trust your proxy. 
```
Some suggestions on shorten and remove repetition.

File: docs/proxy.md [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+* Mount the certificate file into the Envbuilder container and then set `ENVBUILDER_SSL_CERT_FILE=/path/to/cert.pem`.
+* Mount a directory containing all relevant CA certificates into the Envbuilder container and then set `ENVBUILDER_SSL_CERT_DIR=/path/to/certs/`.
+
+## Demonstration
+Envbuilder clones a repository that contains your devcontainer.json and optional Dockerfile so that it can build your container. If the clone is done using HTTPS, then TLS verification will have to succeed, or be disabled. If a transparent HTTPS proxy is present, TLS verification will fail unless Envbuilder trusts the certificate used by the transparent proxy. Therefore, we need to tell Envbuilder how to trust your transparent proxy. 
+
+The summary in the previous section shows how to configure Envbuilder using Terraform for Docker and Kubernetes. For this example we'll use docker directly to avoid complexity that might result in confusion. Docker is also more likely than Terraform to already be installed in your testing environment.
```
Comment (mafredri) [inline docs/proxy.md:22]:
```suggestion
The summary in the previous section shows how to configure Envbuilder for use with proxies. In this example we'll use Docker demonstrate how the configuration can be applied in practice.
```
Nit: This is the first time Terraform is mentioned, so let's just remove that?
Nit: We don't need to make assumptions about their dev environment.

Suggestion: It's a good idea to avoid hinting that, unless dumbed down, the reader might be confused. (One interpretation.)

File: docs/proxy.md [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+...
+#1: 📦 Cloned repository! [711.221369ms]
+```
+
+After some time building, a shell will be presented inside the devcontainer environment specified in envbuilder's own repository. Assuming that envbuilder built and ran successfully, go ahead and exit the container:
+```bash
+exit
+```
+
+Let's now temporarily break Envbuilder by introducing a transparent TLS intercepting proxy. To do this, we'll use [mitmproxy](https://mitmproxy.org/). Start mitmproxy in a container, by running the following:
```
Comment (mafredri) [inline docs/proxy.md:44]:
```suggestion
Let's now temporarily break Envbuilder by introducing a transparent TLS proxy to intercept traffic. To do this, we'll use [mitmproxy](https://mitmproxy.org/). Start mitmproxy in a container by running the following:
```

File: docs/proxy.md [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+docker inspect mitmproxy | jq -r '.[].NetworkSettings.IPAddress'
+```
+yields:
+```
+172.17.0.2
+```
+
+You may see a different IP address. If you do, use that wherever we use `172.17.0.2` below. 
+
+A new directory called certs should also be present in your current working directory. It will contain a CA certificate called  `mitmproxy-ca-cert.pem`. This will be what we provide to Envbuilder to help it trust our proxy.
```
Comment (mafredri) [inline docs/proxy.md:62]:
```suggestion
A new directory called certs should also be present in your current working directory. It will contain a CA certificate called  `mitmproxy-ca-cert.pem`. This will be what we provide to Envbuilder to trust our proxy.
```

File: docs/proxy.md [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+Optionally, inspect the certificates served by mitmproxy:
```
Comment (mafredri) [inline docs/proxy.md:64]: Nit: Should we remove this section? I realize it's optional but it doesn't tell the user why they may need it. It does say notice this and that issues is that. But as a reader, what do I actually do with this information? Why would I care?

File: docs/proxy.md [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    -e ENVBUILDER_INIT_SCRIPT='/bin/sh' \
+    -e ENVBUILDER_GIT_URL='https://github.com/coder/envbuilder.git' \
+    ghcr.io/coder/envbuilder:latest
+```
+From the logs, notice that certificate verification fails. It fails because it doesn't trust the certificate that was provided by mitmproxy:
+```
+Failed to clone repository: clone "https://github.com/coder/envbuilder.git": Get "https://github.com/coder/envbuilder.git/info/refs?service=git-upload-pack": proxyconnect tcp: tls: failed to verify certificate: x509: certificate signed by unknown authority
+```
+
+To fix this, we need to provide a ca certificate that Envbuilder can use to verify the server certificate that mitmproxy serves instead of github's actual certificate. Envbuilder provides a few environment variables to accomplish this. They are documented in the summary section above. For this example, we have the ca certificate saved in a directory. The easiest way to provide the ca certificate to envbuilder is therefore to mount it as a volume in the envbuilder container and tell envbuilder to use it. For this, we can use the `SSL_CERT_FILE` environment variable. The command to run Envbuilder is now:
```
Comment (mafredri) [inline docs/proxy.md:103]:
```suggestion
To resolve this, we need to provide a CA certificate that Envbuilder can use to verify the server certificate served by mitmproxy, instead of GitHub’s. Envbuilder offers environment variables for this, as documented above. In this example, the CA certificate is saved in a directory. The simplest approach is to mount this directory as a volume in the Envbuilder container and use the `SSL_CERT_FILE` environment variable. The command to run Envbuilder is now:
```

Nit: I felt this was a bit wordy so I asked ChatGPT for a revision (above, really basic prompt).

That said, I still feel like there's a bit of repetition with what we've already covered. For example, we mentioned that the CA certificate is saved in a directory and has a specific file name. Instead, we could just refer to that without giving the exact details.

File: docs/proxy.md [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+docker run -it --rm \
+    -v $PWD/certs:/certs \
+    -e SSL_CERT_FILE=/certs/mitmproxy-ca-cert.pem \
+    -e https_proxy=https://172.17.0.2:8080 \
+    -e ENVBUILDER_INIT_SCRIPT='/bin/sh' \
+    -e ENVBUILDER_GIT_URL='https://github.com/coder/envbuilder.git' \
+    ghcr.io/coder/envbuilder:latest
+```
+
+As before, this command should yield a shell inside an Envbuilder built environment. Feel free to test it en then exit the container. Assuming this worked, Envbuilder will have cloned a repository and built the relevant container using a proxy that required a custom ca certificate.
```
Comment (mafredri) [inline docs/proxy.md:114]:
```suggestion
As before, this command should yield a shell inside an Envbuilder built environment. Feel free to test it and then exit the container. Assuming this worked, Envbuilder will have cloned a repository and built the relevant container using a proxy that required accepting a custom CA certificate.
```
Nit: Typo, consistency

File: docs/proxy.md [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+### Bonus
+To prove that Envbuilder did in fact use the proxy, and also because it is interesting to observe, open `http://localhost:8081/` in your local browser and you should see the mitmproxy web interface. In the flow tab, there should be a list of all of the HTTP requests that were required to build the container. The first three will be the requests that were used by git to clone the repository. The rest will be the requests that were used to clone the devcontainer image.
```
Comment (mafredri) [inline docs/proxy.md:117]:
```suggestion
To prove that Envbuilder did in fact use the proxy, and also because it is interesting to observe, open `http://localhost:8081/` in your local browser and you should see the mitmproxy web interface. In the flow tab, there should be a list of all of the HTTP requests that were required to build the container. The first three will be the requests that were used to clone the Git repository. The rest will be the requests that were used to pull the devcontainer image.
```
Nit: There's still a bit of uncertainty here (should) and certainty that the Git operations will only perform three requests (I feel this could change).

**Review by mafredri — COMMENTED** [review]
Nice work on this documentation! There are a few things that I found that I think can be improved, but there's two thoughts I'd like to leave you with that I think can be beneficial to your writing in the future:
- Can this be said in less words, i.e. is this critical to get the point across?
- Can this be made more concrete?

File: docs/proxy.md (hunk 9 above)
Comment (SasSwart) [reply docs/proxy.md:64]: updated the wording to show intent and removed the word optional. it should be more useful now.

File: docs/proxy.md (hunk 12 above)
Comment (SasSwart) [reply docs/proxy.md:117]: updated

File: docs/proxy.md [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+    -e https_proxy=https://172.17.0.2:8080 \
+    -e ENVBUILDER_INIT_SCRIPT='/bin/sh' \
+    -e ENVBUILDER_GIT_URL='https://github.com/coder/envbuilder.git' \
+    ghcr.io/coder/envbuilder:latest
+```
+
+As before, this command yields a shell inside an Envbuilder built environment. Feel free to test it and then exit the container. Assuming this worked, Envbuilder will have cloned a repository and built the relevant container using a proxy that required accepting a custom CA certificate.
+
+### Bonus
+To prove that Envbuilder did in fact use the proxy, and also because it is interesting to observe, open `http://localhost:8081/` in your local browser and you see the mitmproxy web interface. In the flow tab, there be a list of all of the HTTP requests that were required to build the container. The first few requests will be those used to clone the Git repository. The rest will be the requests that were used to pull the devcontainer image.
```
Comment (mafredri) [inline docs/proxy.md:117]:
```suggestion
To prove that Envbuilder did in fact use the proxy, and also because it is interesting to observe, open `http://localhost:8081/` in your local browser and you see the mitmproxy web interface. In the flow tab, there will be a list of all of the HTTP requests that were required to build the container. The first few requests will be those used to clone the Git repository. The rest will be the requests that were used to pull the devcontainer image.
```

**Comment by SasSwart** [comment]
@mafredri I've applied all of your review notes. I'm going ahead and merging this. If you would like to provide further feedback, I will open another PR to update :)

Thanks for the review.

**Review by mafredri — COMMENTED** [review]
Great work on the rewrite, looks good. Spotted a stray deletion without compensation tho 😄

File: docs/proxy.md (hunk 13 above)
Comment (SasSwart) [reply docs/proxy.md:117]: hotfixing
